core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt

```diff
@@ -0,0 +1,756 @@
+    val uf = UnionFind(builder.getTrackSections().size.toInt() * 2)
+    for (nodeIdx in trackNodePool) {
+        val ports = builder.getTrackNodePorts(nodeIdx)
+        var firstEndpoint: EndpointTrackSectionId? = null
+        for (port in ports) {
+            val endpoint = ports[port]
+            if (firstEndpoint == null) {
+                firstEndpoint = endpoint
+            } else {
+                uf.union(firstEndpoint.index.toInt(), endpoint.index.toInt())
+            }
+        }
+    }
+    // union by track-section without detector
+    for (trackSectionIdx in builder.getTrackSections()) {
+        val detectors = builder.getTrackSectionDetectors(trackSectionIdx)
+        if (detectors.size == 0) {
+            val startEndpoint = EndpointTrackSectionId(trackSectionIdx, Endpoint.START)
+            val endEndpoint = EndpointTrackSectionId(trackSectionIdx, Endpoint.END)
+            uf.union(startEndpoint.index.toInt(), endEndpoint.index.toInt())
+        }
+    }
+
+    // now that union-find is complete, create detection zones
+    val rootToZoneMap = mutableMapOf<Int, ZoneId>()
+    for (trackSectionIdx in builder.getTrackSections()) {
+        val detectors = builder.getTrackSectionDetectors(trackSectionIdx)
+
+        // prepare "node" detection zones
+        for (endpoint in Endpoint.entries) {
+            val trackSectionEndpoint = EndpointTrackSectionId(trackSectionIdx, endpoint)
+            val node = builder.getNodeAtEndpoint(trackSectionEndpoint) ?: continue
+
+            val zoneRoot = uf.findRoot(trackSectionEndpoint.index.toInt())
+            if (!rootToZoneMap.containsKey(zoneRoot)) {
+                rootToZoneMap[zoneRoot] = builder.zone(listOf())
+            }
+            val zoneIdx = rootToZoneMap[zoneRoot]!!
+
+            builder.zoneAddNode(zoneIdx, node)
+
+            if (detectors.size != 0) {
+                val dirDetector =
+                    if (endpoint == Endpoint.START) detectors[0].decreasing
+                    else detectors[detectors.size - 1].increasing
+                builder.setNextZone(dirDetector, zoneIdx)
+            }
+        }
+
+        // for each detector pair inside the track section
+        for (detectorIndex in 0 until detectors.size - 1) {
+            val leftDetector = detectors[detectorIndex]
+            val rightDetector = detectors[detectorIndex + 1]
+            val zoneIdx = builder.zone(listOf())
+            builder.setNextZone(leftDetector.increasing, zoneIdx)
+            builder.setNextZone(rightDetector.decreasing, zoneIdx)
+        }
+    }
+}
+
+fun EdgeDirection.toDirection(): Direction {
+    return when (this) {
+        EdgeDirection.START_TO_STOP -> INCREASING
+        EdgeDirection.STOP_TO_START -> DECREASING
+    }
+}
+
+fun EdgeEndpoint.toEndpoint(): Endpoint {
+    return if (this == EdgeEndpoint.BEGIN) Endpoint.START else Endpoint.END
+}
+
+fun parseRJSInfra(rjsInfra: RJSInfra): RawInfra {
```

Comment:
Damn, 430 lines long...

We probably should break it up. Not necessarily in this PR, but we should